src/zarr/registry.py

```diff
@@ -0,0 +1,179 @@
+from __future__ import annotations
+
+import warnings
+from collections import defaultdict
+from typing import TYPE_CHECKING, Any, Generic, TypeVar
+
+if TYPE_CHECKING:
+    from zarr.abc.codec import Codec, CodecPipeline
+    from zarr.buffer import Buffer, NDBuffer
+
+from importlib.metadata import EntryPoint
+from importlib.metadata import entry_points as get_entry_points
+
+from zarr.config import BadConfigError, config
+
+T = TypeVar("T")
+
+
+class Registry(Generic[T], dict[str, type[T]]):
+    def __init__(self) -> None:
+        super().__init__()
+        self.lazy_load_list: list[EntryPoint] = []
+
+    def lazy_load(self) -> None:
+        for e in self.lazy_load_list:
+            self.register(e.load())
+        self.lazy_load_list.clear()
+
+    def register(self, cls: type[T]) -> None:
+        self[fully_qualified_name(cls)] = cls
+
+
+__codec_registries: dict[str, Registry[Codec]] = defaultdict(Registry)
+__pipeline_registry: Registry[CodecPipeline] = Registry()
+__buffer_registry: Registry[Buffer] = Registry()
+__ndbuffer_registry: Registry[NDBuffer] = Registry()
+
+"""
+The registry module is responsible for managing implementations of codecs, pipelines, buffers and ndbuffers and 
+collecting them from entrypoints.
+The implementation used is determined by the config
+"""
+
+
+def _collect_entrypoints() -> list[Registry[Any]]:
+    """
+    Collects codecs, pipelines, buffers and ndbuffers from entrypoints.
+    Allowed syntax for entry_points.txt is e.g.
```

Comment:
```suggestion
    Entry points can either be single items or groups of items.
    Allowed syntax for entry_points.txt is e.g.
```
Why entry_points.txt? Isn't that something that is declared in a pyproject.toml?